pkg/storage/bucket/s3/config.go

```diff
@@ -111,6 +112,7 @@ func (cfg *Config) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
 	f.BoolVar(&cfg.NativeAWSAuthEnabled, prefix+"s3.native-aws-auth-enabled", false, "If enabled, it will use the default authentication methods of the AWS SDK for go based on known environment variables and known AWS config files.")
 	f.Uint64Var(&cfg.PartSize, prefix+"s3.part-size", 0, "The minimum file size in bytes used for multipart uploads. If 0, the value is optimally computed for each object.")
 	f.BoolVar(&cfg.SendContentMd5, prefix+"s3.send-content-md5", false, "If enabled, a Content-MD5 header is sent with S3 Put Object requests. Consumes more resources to compute the MD5, but may improve compatibility with object storage services that do not support checksums.")
+	f.StringVar(&cfg.STSEndpoint, prefix+"s3.sts-endpoint", "", "Accessing S3 resources using temporary, secure credentials provided by AWS Security Token Service")
```

Comment:
I can forsure add help text, for this, and that Endpoint matches with what Minio needs. One question I saw in the docs and I dont know if this is something we need to add but. To enable STS in MinIO Operator, you need to have MinIO Operator 5.0 or later, ensure TLS is configured, and set the OPERATOR_STS_ENABLED environment variable to on.  Do we need to do anything for this?

Reply:
I can for sure add help text, for this, and that Endpoint matches with what Minio needs. One question I saw in the docs and I dont know if this is something we need to add but. To enable STS in MinIO Operator, you need to set the OPERATOR_STS_ENABLED environment variable to on. Do we need to ensure on our side that OPERATOR_STS_ENABLED is set up? I know we use minio to connect to our storages. (s3, azure, gcs).

Reply:
Do we need to add a flag for allowing this behavior and setting it up when we configure minio?

Reply:
never mind, I apologize for this, seems the ENV needs to be configured on the k8 pod themselves, so nothing for us to worry about.

Reply:
> To enable STS in MinIO Operator, you need to set the OPERATOR_STS_ENABLED environment variable to on. Do we need to ensure on our side that OPERATOR_STS_ENABLED is set up? I know we use minio to connect to our storages. (s3, azure, gcs).

My understanding is that the MinIO Operator is for provisioning MinIO clusters in kubernetes. How folks choose to provision minio clusters is outside of Mimir's scope. We also don't use it in the helm chart.

We use the minio library to connect to S3-compatible object store (minio included). But that's purely a protocol layer and from my understanding the setting you added should be sufficient to use STS if it's enabled on the object store (be it minio, aws s3 or different).

> Do we need to add a flag for allowing this behavior and setting it up when we configure minio?

Having said that ☝️ , I don't think we need extra config.

Reply:
This made me realize that the config example you have look a lot like the request that minio accepts for STS. The minio library is the one which adds these parameters to the URL - [code](https://github.com/grafana/mimir/blob/3d62b188ff7a99be0345b8e7e1f721eb7b071a48/vendor/github.com/minio/minio-go/v7/pkg/credentials/sts_web_identity.go#L117-L136). And what it expects is only a hostname with scheme. One example of an [AWS S3 STS endpoint](https://docs.aws.amazon.com/general/latest/gr/sts.html) is `https://sts.us-east-2.amazonaws.com`. It doesn't mention the URL parameters.

It looks like the endpoint should be quite simple. Unless I'm missing something (correct me if I am), then we don't really need elaborate examples. But I would ask to correct the test case to include something more realistic (like a simple URL).

Reply:
I see, looking at docs i thought the client provided these minio request. But it seems you are saying we just need to provide the AWS sts endpoint and minio will set up the correct request itself?

Reply:
> But it seems you are saying we just need to provide the AWS sts endpoint and minio will set up the correct request itself?

yep, that's my understanding

Reply:
I see, you seem correct. Does seem it appends the url values to the aws sts endpoint. Then this brings up a good point. I think we should validate the endpoint 
```
 https://sts.us-east-2.amazonaws.com
```
like we do for the normal aws endpoint. in this code 
```
func (cfg *Config) Validate() error {
	if !util.StringsContain(supportedSignatureVersions, cfg.SignatureVersion) {
		return errUnsupportedSignatureVersion
	}
	if cfg.Endpoint != "" {
		endpoint := strings.Split(cfg.Endpoint, ".")
		if cfg.BucketName != "" && endpoint[0] != "" && endpoint[0] == cfg.BucketName {
			return errInvalidEndpointPrefix
		}
	}
	if !util.StringsContain(supportedStorageClasses, cfg.StorageClass) && cfg.StorageClass != "" {
		return errUnsupportedStorageClass
	}

	return cfg.SSE.Validate()
}
```
since the ` cfg.Endpoint` is ` s3.eu-central-1.amazonaws.com` closely tied to the sts endpoint. Thoughts?

Reply:
i wouldn't impose a lot of validation on the URL. The validation might make sense for AWS S3, but not for other providers like minio. 

Perhaps checking that it's a valid URL should be enough.

Reply:
ah ok, I assumed sts endpoint follow same as `s3.(region).amazonaws.com` but sts intstead. But I can just validate its a valid URL if you think that is okay to do, if not i can skip this.

Reply:
Adding some URL validation would make sense I think. Perhaps to avoid the misunderstanding that just `sts.eu-central-1.amazonaws.com` is not a valid URL, but `https://sts.eu-central-1.amazonaws.com` is. This may be less trivial since golang interprets `sts.eu-central-1.amazonaws.com` as a URL without a host where the path is `sts.eu-central-1.amazonaws.com`. Maybe we can validate that the URL parses as a `net.URL` and that it has a non-empty hostname.

Reply:
yeah that makes sense. I'll make that change.

Reply:
added.